### PR TITLE
pc-tty: cleaned up and reworked implementation

### PR DESCRIPTION
**Pull request by Maxez**
New features:
- BIOS keyboard support,
- extended scan codes,
- new VT100 modes,
- terminal scrolling,
- keyboard LED control

### CODE REVIEW

File: tty/pc-tty/ttypc.c [hunk 1]
```diff
@@ -1,204 +1,184 @@
...
+	case 4:
+		ttypc_vga_destroy(&ttypc_common);
+	case 3:
+		for (j = 0; j < i; j++)
+			ttypc_vt_destroy(ttypc_common.vts + j);
+	case 2:
+		resourceDestroy(ttypc_common.lock);
+	case 1:
+		portDestroy(ttypc_common.port);
+	}
```
Comment (agkaminski) [inline tty/pc-tty/ttypc.c:181]: If we terminate process anyway, there is no real need to destroy resources and unmap memory. OS will do it all for us.

File: tty/pc-tty/ttypc_kbd.c [hunk 2]
```diff
@@ -14,300 +14,388 @@
...
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 119 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 120 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 121 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 122 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 123 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 124 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 125 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 126 */
-	{ KB_NONE,   "",       "",       "",       "",       "" }, /* 127 */
+static keymap_t scodes[] = {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:40]: Maybe static const?

File: tty/pc-tty/ttypc_kbd.c [hunk 3]
```diff
@@ -14,300 +14,388 @@
...
+	{ KB_NONE,   "",        "",        "",        "",        "" },    /* 123 */
+	{ KB_NONE,   "",        "",        "",        "",        "" },    /* 124 */
+	{ KB_NONE,   "",        "",        "",        "",        "" },    /* 125 */
+	{ KB_NONE,   "",        "",        "",        "",        "" },    /* 126 */
+	{ KB_NONE,   "",        "",        "",        "",        "" },    /* 127 */
+};
+
+
+/* KB_BP (keypad keys) modifiers map */
+static unsigned char kpmod[] = {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:174]: Maybe static const?

File: tty/pc-tty/ttypc_kbd.c [hunk 4]
```diff
@@ -14,300 +14,388 @@
...
 };
 
 
-/* Function gets characters from keyboard */
+/* Get key code from keyboard */
 static char *_ttypc_kbd_get(ttypc_t *ttypc)
 {
-	uint8_t dt;
-	char *more = NULL;
+	static char *erreboot = "Failed to reboot\n";
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:197]: static const

File: tty/pc-tty/ttypc_kbd.c [hunk 5]
```diff
@@ -14,300 +14,388 @@
...
-			ttypc->extended = 0;
-			break;
+		/* AutoRepeat Mode */
+		if (!ttypc->vt->arm && (dt == lkey) && (ext == lext))
+			s = NULL;
+
+		/* Reboot sequence - Ctl-Alt-Del */
+		if (s && ext && !strcmp(s, "\033[3~") && (ttypc->shiftst == (KB_CTL | KB_ALT))) {
+			if (reboot(PHOENIX_REBOOT_MAGIC) < 0)
+				ttypc_vt_write(ttypc->vt, 0, erreboot, sizeof(erreboot));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:383]: sizeof(erreboot) = 4 (sizeof(char *)) It should be:
`static const char erreboot[] = "Failed to reboot\n";`
and `ttypc_vt_write ` should expect pointer to const.

File: tty/pc-tty/ttypc_kbd.c [hunk 6]
```diff
@@ -316,62 +404,290 @@ static int ttypc_kbd_interrupt(unsigned int n, void *arg)
...
-	mutexCreate(&ttypc->rlock);
-	condCreate(&ttypc->rcond);
+	return EOK;
+}
+
+
+/* May not work for PS/2 emulation through USB legacy support */
+int _ttypc_kbd_updateled(ttypc_t *ttypc)
+{
+	int err, ret = 0;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:550]: err is only assigned, never used

File: tty/pc-tty/ttypc_kbd.c [hunk 7]
```diff
@@ -316,62 +404,290 @@ static int ttypc_kbd_interrupt(unsigned int n, void *arg)
...
+	/* Attach interrupt */
+	if ((err = interrupt((ttypc->kirq = 1), ttypc_kbd_interrupt, ttypc, ttypc->kcond, &ttypc->kinth)) < 0) {
+		resourceDestroy(ttypc->klock);
+		resourceDestroy(ttypc->kcond);
+		return err;
+	}
+
+	/* Launch keyboard control thread */
+	if ((err = beginthread(ttypc_kbd_ctlthr, 1, &ttypc->kstack, sizeof(ttypc->kstack), (void *)ttypc)) < 0) {
+		resourceDestroy(ttypc->klock);
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_kbd.c:611]: You should do `resourceDestroy` on `ttypc->kinth` too

File: tty/pc-tty/ttypc_vga.c [hunk 8]
```diff
@@ -1,119 +1,258 @@
...
+void _ttypc_vga_switch(ttypc_vt_t *vt)
+{
+	ttypc_t *ttypc = vt->ttypc;
+	ttypc_vt_t *cvt = ttypc->vt;
+
+	if (vt == cvt)
+		return;
+
+	/* VGA memory -> VT memory */
+	memcpy(cvt->mem, cvt->vram, cvt->rows * cvt->cols * CHR_VGA);
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:71]: Side note - pointer to the screen buffer should be of pointer to volatile u16 type.

Comment (agkaminski) [reply tty/pc-tty/ttypc_vga.c:71]: All memory shared between application and hardware should be volatile, really.

File: tty/pc-tty/ttypc_vga.c [hunk 9]
```diff
@@ -1,119 +1,258 @@
...
+	if (vt->scrbpos < vt->rows)
+		memcpy(vt->vram + vt->scrbpos * vt->cols, vt->scro, (vt->rows - vt->scrbpos) * vt->cols * CHR_VGA);
+
+	/* Update cursor */
+	vt->cpos -= n * vt->cols;
+	vt->crow -= n;
+
+	if (vt == vt->ttypc->vt)
+		_ttypc_vga_setcursor(vt);
 }
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:192]: Almost identical code. Maybe `n` should be signed and one function to rule them all?

File: tty/pc-tty/ttypc_vt.c [hunk 10]
```diff
@@ -0,0 +1,737 @@
...
+			}
+			else {
+				vt->lc = 0;
+				vt->ccol++;
+				vt->cpos++;
+			}
+			break;
+
+		case ESC_ESC:
+			switch(c) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:178]: missing space between switch and (c)

File: tty/pc-tty/ttypc_vt.c [hunk 11]
```diff
@@ -0,0 +1,737 @@
...
+				break;
+			}
+			break;
+
+		case ESC_BLANK: /* ESC space */
+			vt->escst = ESC_INIT;
+			break;
+		
+		case ESC_HASH:
+			switch(c) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:335]: missing space between switch and (c)

File: tty/pc-tty/ttypc_vt.c [hunk 12]
```diff
@@ -0,0 +1,737 @@
...
+				break;
+
+			default:
+				vt->escst = ESC_INIT;
+				break;
+			}
+			break;
+
+		case ESC_CSIQM: /* DEC private modes */
+			switch(c)
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:530]: missing space between switch and (c), { shouldn't be in the new line

File: tty/pc-tty/ttypc_vt.c [hunk 13]
```diff
@@ -0,0 +1,737 @@
...
+			break;
+
+		case ESC_DCS:
+			/* TODO */
+			vt->dcsst = DCS_INIT;
+			vt->escst = ESC_INIT;
+			break;
+
+		case ESC_SCA:
+			switch(c) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:587]: missing space between switch and (c)

File: tty/pc-tty/ttypc_vt.c [hunk 14]
```diff
@@ -0,0 +1,737 @@
...
+				break;
+
+			default:
+				vt->escst = ESC_INIT;
+				break;
+			}
+			break;
+
+		case ESC_STR:
+			switch(c) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:599]: missing space between switch and (c)

File: tty/pc-tty/ttypc_vt.c [hunk 15]
```diff
@@ -0,0 +1,737 @@
...
+}
+
+
+ssize_t ttypc_vt_read(ttypc_vt_t *vt, int mode, char *buff, size_t len)
+{
+	return libtty_read(&vt->tty, buff, len, mode);
+}
+
+
+ssize_t ttypc_vt_write(ttypc_vt_t *vt, int mode, char *buff, size_t len)
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:639]: const char *buff

File: tty/pc-tty/ttypc_vtf.c [hunk 16]
```diff
@@ -1,617 +1,639 @@
...
+	BG_LIGHTGREY /* 7 */
+};
 
-	for (i = 0; i < MAXPARMS; i++)
-		virt->parms[i] = 0;
-	virt->parmi = 0;
-}
 
+/* ASCII character set */
+static uint16_t ascii[] = {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:104]: Why do we need table to perform y = x + 0x20 function?

File: tty/pc-tty/ttypc_vtf.c [hunk 17]
```diff
@@ -1,617 +1,639 @@
...
 
-/* device attributes */
-void _ttypc_vtf_da(ttypc_virt_t *virt)
+void _ttypc_vtf_da(ttypc_vt_t *vt)
 {
-#if 0
-	static uint8_t *response = (u_char *)DA_VT220;
+	/* 62 - class 2 terminal, c - end of attributes list */
+	static char *da = "\033[62;c";
+	write(0, da, sizeof(da));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:290]: `sizeof(da) = 4` (`sizeof(char *)`). It should be `static const char da[] = "\033[62;c";`

File: tty/pc-tty/ttypc_vtf.c [hunk 18]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
+	vt->bottom = vt->parms[1] - 1;
+
+	/* Cursor to origin position */
+	vt->cpos = (vt->om) ? vt->top * vt->cols : 0;
+	vt->ccol = 0;
+}
+
+
+/* Apply attr to a VGA character */
+static void ttypc_vtf_putattr(uint16_t *ptr, uint16_t attr) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:762]: `{` should be in the new line

File: tty/pc-tty/ttypc_vtf.c [hunk 19]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
+	/* Keep old bg color if it's different than new fg color */
+	if (((*ptr & 0xf000) >> 4) ^ (attr & 0x0f00))
+		nattr = (*ptr & 0xf000) | (nattr & 0x0fff);
+
+	*ptr = (*ptr & 0x00ff) | nattr;
+}
+
+
+/* Apply attr to VGA character buffer */
+static void ttypc_vtf_applyattr(uint16_t *begin, uint16_t *end, uint16_t attr) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:777]: `{` should be in the new line

File: tty/pc-tty/ttypc_vtf.c [hunk 20]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
+	for (; begin < end; begin++)
+		ttypc_vtf_putattr(begin, attr);
+}
+
+
+/* Apply sgr mode globally */
+static void ttypc_vtf_applysgr(ttypc_vt_t *vt, uint8_t sgr) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:784]: `{` should be in the new line

File: tty/pc-tty/ttypc_vtf.c [hunk 21]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
+		vt->lnm = 0;
+		break;
+	}
+}
+
+
+void _ttypc_vtf_reqtparm(ttypc_vt_t *vt)
+{
+	static char *tparm = "\033[3;1;1;120;120;1;0x";
+	write(0, tparm, sizeof(tparm));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:974]: sizeof(tparm ) = 4 (sizeof(char *)). It should be `static const char tparm [] = "\033[3;1;1;120;120;1;0x";`

File: tty/pc-tty/ttypc.c [hunk 22]
```diff
@@ -1,204 +1,184 @@
...
+		else {
+			if ((err = ttypc_bioskbd_init(&ttypc_common)) < 0)
+				break;
+			ttypc_common.ktype = KBD_BIOS;
+		}
+		printf("pc-tty: Initialized %s keyboard\n", (ttypc_common.ktype) ? "PS/2" : "BIOS");
+		init++;
+
+		/* Start poolthread */
+		if ((err = beginthread(ttypc_poolthr, 1, &ttypc_common.pstack, sizeof(ttypc_common.pstack), (void *)&ttypc_common)) < 0)
```
Comment (agkaminski) [inline tty/pc-tty/ttypc.c:161]: ttypc_common.pstack is already a pointer

File: tty/pc-tty/ttypc_vtf.c [hunk 23]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
 	int i = 0;
 
-	switch (virt->parms[0]) {
-	
-	/* return status */
+	switch (vt->parms[0]) {
+	/* Status */
 	case 5:
-		/* respond(answr); */
+		write(0, stat, sizeof(stat));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:683]: `sizeof(stat) = 4` (`sizeof(char *)`). It should be `static const char stat[]`

File: tty/pc-tty/ttypc_vtf.c [hunk 24]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
+		buff[i++] = '\0';
+
+		write(0, buff, i);
 		break;
 
-	/* return printer status */
+	/* Printer status */
 	case 15:
-		/* respond(panswr); */
+		write(0, print, sizeof(print));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:707]: `sizeof(print) = 4` (`sizeof(char *)`). It should be `static const char print[]`

File: tty/pc-tty/ttypc_vtf.c [hunk 25]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
 		break;
 
-	/* return udk status */
+	/* User Defined Keys status */
 	case 25:
-		/* respond(udkanswr); */
+		write(0, udk, sizeof(udk));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:712]: `sizeof(udk) = 4` (`sizeof(char *)`). It should be `static const char udk[]`

File: tty/pc-tty/ttypc_vtf.c [hunk 26]
```diff
@@ -634,212 +656,340 @@ void _ttypc_vtf_sgr(ttypc_virt_t *virt)
...
 		break;
 
-	/* return language status */
+	/* Language status */
 	case 26:
-		/* respond(langanswr); */
+		write(0, lang, sizeof(lang));
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vtf.c:717]: `sizeof(lang) = 4` (`sizeof(char *)`). It should be `static const char lang[]`

File: tty/pc-tty/ttypc_vtf.c (hunk 16 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:104]: The function below is being used for rendering characters onto the VGA screen using appropriate character set (GL ang GR may point to ascii table or any other character set table). As you can see the general approach is to use a table to resolve characters.
I guess we could use functions (Gs, GL and GR would become function pointers) for resolving characters instead. This way we would get rid of unnecessary tables.
```
/* Writes character to screen buffer */
static void _ttypc_vt_sdraw(ttypc_vt_t *vt, char c)
{
	if ((c >= 0x20) && (c <= 0x7f))
		c = (vt->ss) ? (*vt->Gs)[c - 0x20] : (*vt->GL)[c - 0x20];
	else if (c >= 0xa0)
		c = (*vt->GR)[c - 0xa0];

	*(vt->vram + vt->cpos) = vt->attr | c;
	vt->ss = 0;
}
```

File: tty/pc-tty/ttypc_vga.c (hunk 8 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vga.c:71]: We use memcpy() and memsetw() for reading and writing to VGA memory. These functions discard volatile qualifier, so we would have to write custom functions to preserve it. Should I do it?

File: tty/pc-tty/ttypc_vtf.c (hunk 16 above)
Comment (agkaminski) [reply tty/pc-tty/ttypc_vtf.c]: Ok, this is a valid point imho. No need for change

File: tty/pc-tty/ttypc.c (hunk 1 above)
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c (hunk 2 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c (hunk 3 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c (hunk 4 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c (hunk 5 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c:383]: Fixed.

File: tty/pc-tty/ttypc_vga.c (hunk 8 above)
Comment (agkaminski) [reply tty/pc-tty/ttypc_vga.c:71]: General opinion is that memcpy is no suited for memory mapped devices:
https://stackoverflow.com/questions/36729240/passing-argument-2-of-memcpy-discards-volatile-qualifier-from-pointer-target

I think we need here our own memcpy implementation which operates on volatile buffer.

File: tty/pc-tty/ttypc_kbd.c (hunk 6 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c (hunk 7 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c:608]: Fixed.

File: tty/pc-tty/ttypc_vga.c (hunk 9 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vga.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 10 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 11 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 12 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 13 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 14 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vt.c (hunk 15 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vt.c]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 17 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:290]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 18 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 19 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 20 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 21 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c]: Fixed.

File: tty/pc-tty/ttypc.c (hunk 22 above)
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Fixed.

File: tty/pc-tty/ttypc_bioskbd.c [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+
+int ttypc_bioskbd_init(ttypc_t *ttypc)
+{
+	int err;
+
+	if ((ttypc->kbd = mmap(NULL, _PAGE_SIZE, PROT_READ | PROT_WRITE, 0, OID_PHYSMEM, 0x0)) == MAP_FAILED)
+		return -ENOMEM;
+
+	/* Launch keyboard control thread */
+	if ((err = beginthread(ttypc_bioskbd_ctlthr, 1, &ttypc->kstack, sizeof(ttypc->kstack), (void *)ttypc)) < 0)
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc_bioskbd.c]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 23 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:683]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 24 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:707]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 25 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:712]: Fixed.

File: tty/pc-tty/ttypc_vtf.c (hunk 26 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_vtf.c:717]: Fixed.

**Comment by kemonats** [comment]
I found some trailing whitespaces in your code:
```
tty/pc-tty/ttypc.h:1: trailing whitespace.
tty/pc-tty/ttypc_kbd.h:1: trailing whitespace.
tty/pc-tty/ttypc_vga.c:1: trailing whitespace.
tty/pc-tty/ttypc_vga.h:1: trailing whitespace.
tty/pc-tty/ttypc_vt.c:1: trailing whitespace.
tty/pc-tty/ttypc_vt.c:125: trailing whitespace.
tty/pc-tty/ttypc_vt.c:331: trailing whitespace.
tty/pc-tty/ttypc_vt.h:1: trailing whitespace.
tty/pc-tty/ttypc_vt.h:138: trailing whitespace.
tty/pc-tty/ttypc_vtf.c:1: trailing whitespace.
tty/pc-tty/ttypc_vtf.c:450: trailing whitespace.
tty/pc-tty/ttypc_vtf.c:504: trailing whitespace.
tty/pc-tty/ttypc_vtf.c:665: trailing whitespace.
tty/pc-tty/ttypc_vtf.c:690: trailing whitespace.
tty/pc-tty/ttypc_vtf.h:1: trailing whitespace.
```
Could you correct that.

**Comment by Maxez** [comment]
> I found some trailing whitespaces in your code:
> 
> ```
> tty/pc-tty/ttypc.h:1: trailing whitespace.
> tty/pc-tty/ttypc_kbd.h:1: trailing whitespace.
> tty/pc-tty/ttypc_vga.c:1: trailing whitespace.
> tty/pc-tty/ttypc_vga.h:1: trailing whitespace.
> tty/pc-tty/ttypc_vt.c:1: trailing whitespace.
> tty/pc-tty/ttypc_vt.c:125: trailing whitespace.
> tty/pc-tty/ttypc_vt.c:331: trailing whitespace.
> tty/pc-tty/ttypc_vt.h:1: trailing whitespace.
> tty/pc-tty/ttypc_vt.h:138: trailing whitespace.
> tty/pc-tty/ttypc_vtf.c:1: trailing whitespace.
> tty/pc-tty/ttypc_vtf.c:450: trailing whitespace.
> tty/pc-tty/ttypc_vtf.c:504: trailing whitespace.
> tty/pc-tty/ttypc_vtf.c:665: trailing whitespace.
> tty/pc-tty/ttypc_vtf.c:690: trailing whitespace.
> tty/pc-tty/ttypc_vtf.h:1: trailing whitespace.
> ```
> 
> Could you correct that.

Sure, I've removed the trailing whitespaces in the very next commit.

**Comment by Maxez** [comment]
> Are you ready to merge changes?

Yes, the PR is ready for merging. I have no other changes.

**Pull request merged**